### DISCUSSION TITLE
Can I make formatting opt-in per file with `fmt: on`?

### DISCUSSION THREAD

**Discussion by jwhitaker-gridcog** [astral-sh/ruff#15205] [Q&A]
Hey guys,
We have a big company codebase that we're wanting to introduce formatting to. It's data-science heavy, so there's a lot of pandas calls that ruff doesn't handle so well, we'd like to leave them alone. However we'd like to enable formatting for other more standard python files. We'd like this to be opt-in instead of opt-out, however.

Is it possible to configure ruff so that files are considered `fmt: off` by default, but then we could include `# fmt: on` ourselves as an initial line of files we want to opt in to?

This would be a better workflow that the existing exclude list just because the list of files is so damn big, and our in/out status doesn't really generalize by parent paths.

Cheers!

**Comment by dhruvmanila** [comment]
You could use globs in [`ruff.format.exclude`](https://docs.astral.sh/ruff/settings/#format_exclude) to exclude entire folders from being formatted and then remove one folder at a time to gradually introduce the formatter to the codebase. Refer to the [`globset` documentation](https://docs.rs/globset/latest/globset/#syntax) for the syntax.

> It's data-science heavy, so there's a lot of pandas calls that ruff doesn't handle so well, we'd like to leave them alone.

If you don't mind, I'm curious to know what kind of calls are we talking about here. Is it related to https://github.com/astral-sh/ruff/discussions/8452 ?

**Reply by jwhitaker-gridcog** [reply]
thanks for the response - as said I was hopeful to not rely on the exclude list, but appreciate the pointer still.
As an aside the DS-related calls that format poorly are
1) chained calls, issue is https://github.com/astral-sh/ruff/issues/8598 .
2) calls with an obnoxious amount of arguments: we format these with multiple per line, split over multiple lines, to keep line breaks to a minimum. Example:
    ```py
    some_object = {
        'arg_1': 'value_1', 'arg_2': 'value_2', 'arg_3': 'value_3', 'arg_4': 'value_4', 'arg_5': 'value_5',
        'arg_6': 'value_6', 'arg_7': 'value_7', 'arg_8': 'value_8', 'arg_9': 'value_9', ...
    }
    ```

**Reply by dhruvmanila** [reply]
> thanks for the response - as said I was hopeful to not rely on the exclude list, but appreciate the pointer still.

Apologies, I did read you mention about that but I was under the impression that you don't want to add individual _files_. For this reason, I suggested using globs instead to avoid specifying long list of files.

Unfortunately, I don't think we have a way to include files specifically for the formatter so the opt-in behavior you're looking for is unavailable. It might be something that we look into when working on https://github.com/astral-sh/ruff/issues/8232.
